### DISCUSSION TITLE
How to deal correctly with the deprecation of `Column.copy()`?

### DISCUSSION THREAD

**Discussion by dwt** [sqlalchemy/sqlalchemy#9696] [Usage Questions] [answered]
I noticed this deprecation of `Column.copy()` in my unit tests a while ago and am now at the point where I want to deal with it. So... what to use instead?

What do I use it for? We build translation tables with a decorator. So the code looks something like this currently (cruelly shortened for brevity):

```python

@translatable('title', 'description', 'rules')
class Game(Base, MetadataMixin, TranslatableMixin):
    __tablename__ = 'game'
    id = Column(Integer, primary_key=True)
    title = Column(UnicodeText, default=u'')
    description = Column(UnicodeText, default=u'')
    rules = Column(UnicodeText, default=u'')

@translation_for(Game)
class GameTranslation(object):
    __tablename__ = 'game_translation'
```

and the decorator looks roughly like this:

```python
def translatable(*translatable_properties):
    def decorator(a_translatable_class):
        a_translatable_class._translatable_properties = translatable_properties
        return a_translatable_class
    
    return decorator

def translation_for(class_to_translate):
    def translation_subclass_creator(cls):
        bookkeeping_properties = dict(
            id=Column(Integer, primary_key=True),
            language=Column(Unicode(200), nullable=False),
            translation_for_id=Column(Integer,
                ForeignKey(class_to_translate.id, onupdate="CASCADE", ondelete="CASCADE")),
        )
        
        class TranslationForSuper(object):
            pass
        
        new_class_dict = dict()
        new_class_dict.update(bookkeeping_properties)
        for property_name in class_to_translate._translatable_properties:
            new_class_dict[property_name] = getattr(class_to_translate, property_name).property.columns[0].copy()
        new_class_dict['__module__'] = class_to_translate.__module__
        
        translation_model = type(cls.__name__, (cls, TranslationForSuper, Base, MetadataMixin), new_class_dict)
        class_to_translate._model_for_translations = translation_model
        return translation_model
    
    return translation_subclass_creator
````

The key point here being the copying of the attributes here:

```python
        for property_name in class_to_translate._translatable_properties:
            new_class_dict[property_name] = getattr(class_to_translate, property_name).property.columns[0].copy()
```

How would I replicate this functionality without copy? What would be the right way to do this?

**Comment by dwt** [comment]
:/ my search foo has obviously left me, only now I found some other threads about the same topic - recommended solutions there where to go to a mixin for the common columns or use `Table.to_metadata()` to get the columns. That's what I'm lookin into now, but I would still appreciate some guidance.

**Reply by zzzeek** [reply, marked as answer]
A "translateable" column seems to me that it would always be:

1. UnicodeText
2. have no foreign key constraints
3. maybe nullable or not null
4. have no CHECK constraints
5. have no server or client side default generators (OK it looks like you have a client side default gen for some of them, this is still simple)

Why not simply create a new Column object with that very minimal set of parameters?   The reason column._copy() is not public is that there's a lot of constraints and defaults that may have dependencies on the table itself and it is not really feasible to just "copy" these

**Comment by dwt** [comment]
That's actually a really good point. I'll check if there is anything special with these columns, but you are right, the most I can think of are some length constraints. I'll look into this.

**Comment by dwt** [comment]
This worked beautifully. Thanks a bunch!

This is what I resorted to:

```python
        def translatable_property_for(a_property_name):
            column_to_translate = getattr(class_to_translate, a_property_name).property.columns[0]
            assert isinstance(column_to_translate, Column)
            assert isinstance(column_to_translate.type, (UnicodeText, Unicode))
            return Column(
                column_to_translate.type,
                nullable=column_to_translate.nullable,
                default=column_to_translate.default
            )

...
        for property_name in class_to_translate._translatable_properties:
            new_class_dict[property_name] = translatable_property_for(property_name)
```
